plugin: add EndpointResolver interface

thaJeztah opened the pull request:
related to https://github.com/moby/moby/pull/43298#discussion_r820972861

This defines the interface that the package expects in order to lookup pull endpoints, instead of requiring the whole registry.Service interface

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**

thaJeztah committed:
plugin: add EndpointResolver interface

This defines the interface that the package expects in order to lookup
pull endpoints, instead of requiring the whole registry.Service interface.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
Oh! Hmm.. looks like I screwed something up. Will check when back at my computer 

```
INFO [runner] linters took 27.968139049s with stages: goanalysis_metalinter: 27.887976013s 
[2022-03-11T14:42:36.808Z] plugin/manager.go:98:18: SA5007: infinite recursive call (staticcheck)
[2022-03-11T14:42:36.808Z] 	repoInfo, err = s.ResolveRepository(name)
[2022-03-11T14:42:36.808Z] 	                ^
```

cpuguy83 commented on plugin/manager.go:
```diff
@@ -86,11 +91,11 @@ type controller struct {
 
 func (s pluginRegistryService) ResolveRepository(name reference.Named) (repoInfo *registry.RepositoryInfo, err error) {
-	repoInfo, err = s.Service.ResolveRepository(name)
+	repoInfo, err = s.ResolveRepository(name)
```
This is recursive? Also I don't think anything on `pluginRegistryService` implements `ResolveRepository` except itself.

thaJeztah replied:
Yes, I was coming to the same conclusion; just pushed a commit that removes the whole `pluginRegistryService` type

thaJeztah committed:
plugin: remove unused pluginRegistryService

It wrapped the regular registry service, but the ResolveRepository() function
was not called anywhere.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
@cpuguy83 @corhere all green now; ptal

corhere reviewed (approved):
LGTM!

cpuguy83 commented on plugin/manager.go:
```diff
@@ -83,25 +87,8 @@ type controller struct {
-func (s pluginRegistryService) ResolveRepository(name reference.Named) (repoInfo *registry.RepositoryInfo, err error) {
-	repoInfo, err = s.Service.ResolveRepository(name)
-	if repoInfo != nil {
-		repoInfo.Class = "plugin"
```
Is this being set somewhere else?

thaJeztah replied:
That's what I was wondering; this code didn't appear to be called though, so not by this code

thaJeztah replied:
I could double check (not near a computer), but at least couldn't find where it was called

thaJeztah replied:
Think that could relate to the failures on https://github.com/moby/moby/pull/42889 ?

cpuguy83 reviewed (approved):
LGTM
